### PR TITLE
Android WebView Java Bridge error fix

### PR DESCRIPTION
**Pull request by webdevbyjoss**
This fixes the issue in the event emitter with lost context for the Android webview.

## Description of this change
That is a JavaScript binding issue, where the callback we are trying to call looses the this context.
See more details in this thread: 
https://github.com/react-native-webview/react-native-webview/issues/323#issuecomment-511824940
Lets just call the corresponding method directly, without assigning it to local variable.

**Also suggest we log missing handlers as just warnings, we don’t have to register all of them anyway.** 

## Why is this change being made?
This should enable events to be fired for Adnroid webview

## How was this tested? How can the reviewer verify your testing?
It can be tested with Android WebView sample app.
https://github.com/ketch-sdk/ketch-samples/pull/22

Steps to reproduce:
1. Run the app in the Android emulator
2. Go to  chrome://inspect in your browser
3. Observe the events fired in the console

No crashes or any other errors should be visible.

## Checklist

- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I have evaluated the security impact of this change, and [OWASP Secure Coding Practices](https://owasp.org/www-project-secure-coding-practices-quick-reference-guide/migrated_content#) have been observed.
- [x] I have informed stakeholders of my changes.

### CODE REVIEW

File: src/ketch.ts [hunk 1]
```diff
@@ -1309,18 +1309,16 @@ export class Ketch extends EventEmitter {
    */
   emit(eventName: string | symbol, ...args: any[]): boolean {
     if (window.androidListener) {
```
Comment (sethyates) [inline src/ketch.ts:1311]: Change the argument to `event and then add `const eventName = event.toString()` then just use `eventName` everywhere below.

File: src/ketch.ts [hunk 2]
```diff
@@ -1309,18 +1309,16 @@ export class Ketch extends EventEmitter {
...
-      const listener = window.androidListener[eventName.toString()]
-      if (listener) {
-        listener(JSON.stringify(args))
+      if (eventName.toString() in window.androidListener) {
+        window.androidListener[eventName.toString()](JSON.stringify(args))
```
Comment (sethyates) [inline src/ketch.ts]: We should probably be a bit smarter with `args` here too. Check if args is empty, call the handler without args. If the args has a single argument, stringify that. If the args are multiple, stringify as we do now. Same applies for iOS below.

File: src/ketch.ts [hunk 3]
```diff
@@ -1307,23 +1307,22 @@ export class Ketch extends EventEmitter {
...
+      if (eventName in window.webkit.messageHandlers) {
+        window.webkit.messageHandlers[eventName].postMessage(JSON.stringify(args))
       } else {
-        console.error(`Can't pass message to native code because ${eventName.toString()} handler is not registered`)
+        console.warn(`Can't pass message to native code because ${eventName} handler is not registered`)
       }
     }
 
-    return super.emit(eventName, ...args)
+    return super.emit(event, ...args)
```
Comment (sethyates) [inline src/ketch.ts:1326]: This one should still be `eventName`.

File: src/ketch.ts (hunk 2 above)
Comment (webdevbyjoss) [reply src/ketch.ts]: good catch

File: src/ketch.ts [hunk 4]
```diff
@@ -1307,23 +1307,26 @@ export class Ketch extends EventEmitter {
...
     if (window.androidListener) {
-      const listener = window.androidListener[eventName.toString()]
-      if (listener) {
-        listener(JSON.stringify(args))
+      if (eventName in window.androidListener) {
+        if (args.length === 0) {
+          window.androidListener[eventName](JSON.stringify(args))
+        } else {
+          window.androidListener[eventName]()
+        }
```
Comment (sethyates) [inline src/ketch.ts:1318]: Logic is wrong.

```
if (args.length === 0) {
  window.androidListener[eventName]()
} else if (args.length === 1) {
  window.androidListener[eventName](args[0])
} else {
  window.androidListener[eventName](JSON.stringify(args))
}
```

File: src/ketch.ts [hunk 5]
```diff
@@ -1307,23 +1307,26 @@ export class Ketch extends EventEmitter {
...
       } else {
-        console.error(`Can't pass message to native code because ${eventName.toString()} handler is not registered`)
+        console.warn(`Can't pass message to native code because ${eventName} handler is not registered`)
       }
     } else if (window.webkit?.messageHandlers) {
-      const listener = window.webkit.messageHandlers[eventName.toString()]
-      if (listener) {
-        listener.postMessage(JSON.stringify(args))
+      if (eventName in window.webkit.messageHandlers) {
+        window.webkit.messageHandlers[eventName].postMessage(JSON.stringify(args))
```
Comment (sethyates) [inline src/ketch.ts:1324]: Logic is wrong.

```
if (args.length === 0) {
  window.webkit.messageHandlers[eventName]()
} else if (args.length === 1) {
  window.webkit.messageHandlers[eventName](args[0])
} else {
  window.webkit.messageHandlers[eventName](JSON.stringify(args))
}
```

File: src/ketch.ts [hunk 6]
```diff
@@ -1307,23 +1308,28 @@ export class Ketch extends EventEmitter {
...
-  emit(eventName: string | symbol, ...args: any[]): boolean {
+  emit(event: string | symbol, ...args: any[]): boolean {
+    const eventName = event.toString()
     if (window.androidListener) {
-      const listener = window.androidListener[eventName.toString()]
-      if (listener) {
-        listener(JSON.stringify(args))
+      if (eventName in window.androidListener) {
+        if (args.length === 0) {
+          window.webkit.messageHandlers[eventName]()
```
Comment (sethyates) [inline src/ketch.ts:1316]: `window.androidListener`

File: src/ketch.ts [hunk 7]
```diff
@@ -1307,23 +1308,28 @@ export class Ketch extends EventEmitter {
...
+        } else if (args.length === 1) {
+          window.webkit.messageHandlers[eventName](args[0])
```
Comment (sethyates) [inline src/ketch.ts:1318]: Do we always want to stringify if not a string?

File: src/ketch.ts [hunk 8]
```diff
@@ -1307,23 +1308,28 @@ export class Ketch extends EventEmitter {
...
+    const eventName = event.toString()
     if (window.androidListener) {
-      const listener = window.androidListener[eventName.toString()]
-      if (listener) {
-        listener(JSON.stringify(args))
+      if (eventName in window.androidListener) {
+        if (args.length === 0) {
+          window.androidListener[eventName]()
+        } else if (args.length === 1) {
+          window.androidListener[eventName](args[0])
```
Comment (sethyates) [inline src/ketch.ts:1318]: Should we always stringify if not a string?

**Review by sethyates — APPROVED** [review]

**Pull request merged**